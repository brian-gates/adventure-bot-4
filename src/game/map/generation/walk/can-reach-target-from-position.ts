import type { Position } from "~/game/map/index.ts";

/**
 * Determine if the target can be reached from the position by moving
 * diagonally.
 */
export function canReachTargetFromPosition({
  target,
  position,
}: {
  target: Position;
  position: Position;
}) {
  const xDistance = Math.abs(position.col - target.col);
  const yDistance = Math.abs(position.row - target.row);
  return xDistance <= yDistance;
}
