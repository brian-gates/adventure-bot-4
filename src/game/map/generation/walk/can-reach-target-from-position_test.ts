import { assertSnapshot } from "https://deno.land/std@0.224.0/testing/snapshot.ts";
import type { Map, Position } from "~/game/map/index.ts";
import { canReachTargetFromPosition } from "./can-reach-target-from-position.ts";

const start = { row: 0, col: 3 };
const boss = { row: 14, col: 3 };

Deno.test("canReachTargetFromPosition to boss", (t) => {
  assertSnapshot(
    t,
    gridMap({
      map: createMap({
        rows: 10,
        cols: 10,
      }),
      target: boss,
    }),
  );
});

Deno.test("canReachTargetFromPosition to start", (t) => {
  assertSnapshot(
    t,
    gridMap({
      map: createMap({
        rows: 10,
        cols: 10,
      }),
      target: start,
    }),
  );
});

Deno.test("canReachTargetFromPosition: 10x10", (t) => {
  assertSnapshot(
    t,
    gridMap({
      map: createMap({
        rows: 10,
        cols: 10,
      }),
      target: {
        row: 5,
        col: 5,
      },
    }),
  );
});

Deno.test("canReachTargetFromPosition: 10x10 reverse", (t) => {
  const map = createMap({
    rows: 10,
    cols: 10,
  });
  assertSnapshot(t, gridMap({ map, target: start }));
});

function createMap({
  rows = 15,
  cols = 7,
  locations = [],
  paths = [],
  guild = null,
  id = "test-map",
  createdAt = new Date(),
  updatedAt = new Date(),
  guildId = null,
}: Partial<Map> = {}): Map {
  return {
    rows,
    cols,
    locations,
    paths,
    guild,
    id,
    createdAt,
    updatedAt,
    guildId,
  };
}

function gridMap({
  map,
  target,
}: {
  map: Map;
  target: Position;
}) {
  let output = "";
  for (let row = 0; row < map.rows; row++) {
    let line = "";
    for (let col = 0; col < map.cols; col++) {
      const valid = canReachTargetFromPosition({
        target,
        position: { row, col },
      });
      line += valid ? "O" : ".";
    }
    output += line + "\n";
  }
  return output;
}
