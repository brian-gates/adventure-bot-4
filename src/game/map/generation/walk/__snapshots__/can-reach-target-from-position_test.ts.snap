export const snapshot = {};

snapshot[`canReachTargetFromPosition to boss 1`] = `
"OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOO.
"
`;

snapshot[`canReachTargetFromPosition to start 1`] = `
"...O......
..OOO.....
.OOOOO....
OOOOOOO...
OOOOOOOO..
OOOOOOOOO.
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
"
`;

snapshot[`canReachTargetFromPosition: 10x10 1`] = `
"OOOOOOOOOO
.OOOOOOOOO
..OOOOOOO.
...OOOOO..
....OOO...
.....O....
....OOO...
...OOOOO..
..OOOOOOO.
.OOOOOOOOO
"
`;

snapshot[`canReachTargetFromPosition: 10x10 reverse 1`] = `
"...O......
..OOO.....
.OOOOO....
OOOOOOO...
OOOOOOOO..
OOOOOOOOO.
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
OOOOOOOOOO
"
`;
